### ISSUE TITLE
Pass triggering event on ContextMenu.Trigger to event handler

### ISSUE THREAD

**Issue by tmcw** [radix-ui/primitives#766]
### Overview

My application includes a map, which can be right-clicked in order to bring up more details. The map is rendered with WebGL, so is basically opaque to React. I'd like to be able to use the ContextMenu component and show more details about what was right-clicked. To make this possible, it would be nice to pass the event that triggered the context menu open to a method that looks up map details.

### Who does this impact? Who is this for?

This will be nice if any of the things that trigger a context menu are, well, contextual to what was right-clicked. This map example is certainly one, but something like having one context menu at a top element and using, basically, delegation to trigger different right-click menus, would be an example.

### Additional context

API might look like:

```tsx
<Root onOpenChange={((open: boolean, event: React.MouseEvent | React.PointerEvent))}>
```

**Comment by benoitgrelard** [comment]
Hey @tmcw, thanks for opening an issue.
In the case of your map example, can I ask how you're planning to decide what to show then?
I'm curious cause you say it's basically a webGL thing, opaque to React. Would you be working out what to show based on the pointer position in the event?

**Comment by jjenzz** [comment]
> I'd like to be able to use the ContextMenu component and show more details about what was right-clicked

If you're wanting to render **information** about something then it's probably not the appropriate use of the `ContextMenu` component [which is meant for menus](https://en.wikipedia.org/wiki/Context_menu).

The [`Popover`](https://www.radix-ui.com/docs/primitives/components/popover) can be controlled and might be more suited to your use-case?

> [..] but something like having one context menu at a top element and using, basically, delegation to trigger different right-click menus, would be an example

For this, your app can be a `<ContextMenu.Trigger as="div" />` and you can nest `ContextMenu`s inside.

**Comment by tmcw** [comment]
> I'm curious cause you say it's basically a webGL thing, opaque to React. Would you be working out what to show based on the pointer position in the event?

Yep, exactly this. In particular, using Mapbox, which gives you a [method to query features at a particular point within the WebGL canvas](https://docs.mapbox.com/mapbox-gl-js/api/map/#map#queryrenderedfeatures), based on the offset from the top-left corner.

> If you're wanting to render information about something then it's probably not the appropriate use of the ContextMenu component which is meant for menus.

This will be a menu - basically, think of it as the same as the right-click menu on Figma. For example: right click on a stack of overlapping features and choose which feature to select.

**Comment by benoitgrelard** [comment]
@tmcw the reason I asked is because I know we have a similar instance in our product (Modulz) where we haven't been able to use the primitive for that reason so it's definitely of interest for us to provide a way to make this work somehow.

I tested this solution quickly yesterday and it is workable. However, I am not sure that it is the best solution semantically. We may be better off raising a custom event.

The reason for this is that, if it wasn't for touch support, you could already do what you want:
- wrap your canvas in a `ContextMenu.Trigger`
- attach your own `onContextMenu` event to get access to the point coordinates in time

However, because we support touch via a long-press, it wouldn't be straightforward for you to hook into that one or re-create it. So I'm thinking perhaps we raise a custom event/callback for when we generally detect an event (context menu or long press) that should open the menu.

**Comment by jjenzz** [comment]
> So I'm thinking perhaps we raise a custom event/callback for when we generally detect an event (context menu or long press) that should open the menu.

We already have an `onOpenChange` that triggers when that happens so it sounds like the combination of the two is what is needed?

```jsx
export default () => {
  const pointRef = React.useRef<number[] | null>(null);

  return (
    <ContextMenu.Root
      onOpenChange={(open) => {
        if (open) {
          queryRenderedFeatures(pointRef.current);
        } else {
          pointRef.current = null;
		}
      }}
    >
      <ContextMenu.Trigger
        as="canvas"
        onContextMenu={(event) => (pointRef.current = [event.clientX, event.clientY])}
      />
      <ContextMenu.Content>
        <ContextMenu.Item>One</ContextMenu.Item>
        <ContextMenu.Item>Two</ContextMenu.Item>
        <ContextMenu.Item>Three</ContextMenu.Item>
      </ContextMenu.Content>
    </ContextMenu.Root>
  );
}
```

**Comment by benoitgrelard** [comment]
That wouldn't work with the long press event though, like I was saying above.

**Comment by jjenzz** [comment]
I think the fact that I used `onContextMenu` has confused things.

The `onOpenChange` will fire when the menu is opened (when long press has occurred) which is where I look up the map details. My thinking is that the snippet just needs updating to get the `point` using whichever events @tmcw wants to support... it could be `onMouseDown`/`onPointerDown` instead if preferred. 

Unfortunately though, I just tried `onPointerDown` to demonstrate and it doesn't get the point in time if the menu is already open 🤔 https://codesandbox.io/s/weathered-bash-2ze4e

**Comment by benoitgrelard** [comment]
Yeah I mean technically you could do it like that, perhaps even always update the point ref on `pointermove` for example.

What I'm trying to get to is that I think semantically it's abusing the `onOpenChange` callback a bit.
It only works because `ContextMenu` cannot be controlled but if it could, the menu would have other ways of opening potentially.

That's why I suggested potentially a custom event because, the information needed is rather about where the point was in the interaction that triggered the menu to open. If all browsers worked properly with only `onContextMenu` this is how you'd do it, so it seems to me it should still be done the same way.

**Comment by jjenzz** [comment]
Why would it matter if it was controlled? My immediate thought when I saw this ticket was to control it, but obviously we don't allow that atm. 

> the information needed is rather about where the point was in the interaction that triggered the menu to open

I'm curious what the new event you're thinking of would look like? `onPointerOpen`? The things is, we should already be able to get this information without something like that because `onPointerDown` on trigger + `open === true` is `onPointerOpen`.

I'd favour allowing it to be controlled and figuring out why `onPointerDown` event doesn't fire when the menu re-opens. Otherwise, it feels we'll end up going off-piste in regards to consistency. Where do we draw the line? What if someone needs to know if it was a long press that opened it, do we provide a longpress bool as part of our API? 

The coords requirement is very specific to @tmcw's product (because of taking things out of React-land) and that is what controlling things is for in my mind. 

However, I do wonder if it is worth considering the `CustomEvent` idea more seriously that we spoke about a while ago (where all our callbacks would be `CustomEvent`s). That would allow us to provide `event.detail` with information like this without exposing implementation detail (the native event).

```jsx
onOpenChange={event => {
	console.log(event.detail) // { open: true, triggerPoint: null } - trigger wasn't clicked
	console.log(event.detail) // { open: true, triggerPoint: [12, 100] } - trigger was clicked
	event.preventDefault() // doesn't prevent anything native
}}
```

**Comment by SimeonC** [comment]
I'm interested in this kind of implementation too so I thought I'd outline my use case here to add some fuel to the fire.
I have a "Layout Editor" component using SVGs to build a floorplan and wanted to use a context menu to allow users to add elements at a particular place. (`Right-click > Add Shape`)

This is workable currently in a rather roundabout way (though in my case this is more of how I wanted it to work). I have the following DOM structure. My way of detecting "what" to show in the menu is kind of a bit custom in that I handle the "mousedown" of the `svg` and then figure out what was clicked to then populate the context menu. Radix ContextMenu places it in the correct place for here so all I needed was to be able to figure out "what" was clicked. With the canvas you may be able to use the same approach.
```<ContextMenu.Root>
  <ContextMenu.Content>...</ContextMenu.Content>
  <ContextMenu.Trigger>
    <svg onMouseDown={setTarget}>...</svg>
  <ContextMenu.Trigger>
</ContextMenu.Root>
```

**Comment by benoitgrelard** [comment]
Closing this as we are tracking in #1307
